arangod/Aql/IResearchViewExecutor.cpp

```diff
@@ -0,0 +1,591 @@
+          arangodb::iresearch::DocumentPrimaryKey::read(documentId, key);
+
+      TRI_ASSERT(readSuccess == documentId.isSet());
+
+      if (!readSuccess) {
+        LOG_TOPIC("6442f", WARN, arangodb::iresearch::TOPIC)
+            << "failed to read document primary key while reading document "
+               "from arangosearch view, doc_id '"
+            << docId << "'";
+      }
+    }
+
+    return true;
+  } else {
+    documentId = {};
+  }
+
+  return false;
+}
+
+template <bool ordered>
+bool IResearchViewExecutor<ordered>::writeRow(ReadContext& ctx, IndexReadBufferEntry bufferEntry) {
+  LocalDocumentId const& documentId = _indexReadBuffer.getId(bufferEntry);
+  TRI_voc_cid_t const& cid = _indexReadBuffer.getCid();
+
+  TRI_ASSERT(documentId.isSet());
+
+  Query& query = infos().getQuery();
+  auto collection = lookupCollection(*query.trx(), cid, query);
+
+  if (!collection) {
+    std::stringstream msg;
+    msg << "failed to find collection while reading document from "
+           "arangosearch view, cid '"
+        << cid << "'";
+    query.registerWarning(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND, msg.str());
+    return false;
+  }
+
+  // read document from underlying storage engine, if we got an id
+  if (collection->readDocumentWithCallback(query.trx(), documentId, ctx.callback)) {
+    // in the ordered case we have to write scores as well as a document
+    if /* constexpr */ (ordered) {
+      // scorer register are placed consecutively after the document output register
+      RegisterId scoreReg = ctx.docOutReg + 1;
+
+      for (auto& it : _indexReadBuffer.getScores(bufferEntry)) {
+        TRI_ASSERT(infos().isScoreReg(scoreReg));
+        bool mustDestroy = false;
+        AqlValueGuard guard{it, mustDestroy};
+        ctx.outputRow.moveValueInto(scoreReg, ctx.inputRow, guard);
+        ++scoreReg;
+      }
+
+      // we should have written exactly all score registers by now
+      TRI_ASSERT(!infos().isScoreReg(scoreReg));
+    }
+
+    return true;
+  }
+
+  return false;
+}
+
+template <bool ordered>
+void IResearchViewExecutor<ordered>::evaluateScores(ReadContext& ctx) {
+  // This must not be called in the unordered case.
+  TRI_ASSERT(ordered);
+
+  // evaluate scores
+  _scr->evaluate();
+
+  // in arangodb we assume all scorers return float_t
+  auto begin = reinterpret_cast<const float_t*>(_scrVal.begin());
+  auto end = reinterpret_cast<const float_t*>(_scrVal.end());
+
+  // scorer register are placed consecutively after the document output register
+  // is used here currently only for assertions.
+  RegisterId scoreReg = ctx.docOutReg + 1;
+
+  // copy scores, registerId's are sequential
+  for (; begin != end; ++begin, ++scoreReg) {
+    TRI_ASSERT(infos().isScoreReg(scoreReg));
+    _indexReadBuffer.pushScore(*begin);
+  }
+
+  // We should either have no _scrVals to evaluate, or all
+  TRI_ASSERT(begin == nullptr || !infos().isScoreReg(scoreReg));
+
+  while (infos().isScoreReg(scoreReg)) {
```

Comment:
is that possible?

Reply:
It occurs at least during the catch tests, because `_scrVal.begin()` is sometimes `nullptr`. If you think this is strange, we should investigate.